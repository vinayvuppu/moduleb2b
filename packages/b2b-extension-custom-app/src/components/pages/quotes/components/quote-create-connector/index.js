export { default } from './quote-create-connector';
