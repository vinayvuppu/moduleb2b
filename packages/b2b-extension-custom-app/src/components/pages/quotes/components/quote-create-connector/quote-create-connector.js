import { useContext, useState } from 'react';
import { useLocation } from 'react-router-dom';
import { useMutation } from 'react-apollo';
import PropTypes from 'prop-types';

import UpdateQuote from '../../graphql/UpdateQuote.graphql';
import CreateQuote from '../../graphql/CreateQuote.graphql';

import B2BApolloClientContext from '../../../../common/b2b-apollo-client-context';

import { QUOTE_TYPES } from '../../constants';

export const QuoteCreateConnector = props => {
  const b2bApolloClientContext = useContext(B2BApolloClientContext.Context);
  const location = useLocation();
  const [quote, setQuote] = useState();
  const [isLoading, setIsLoading] = useState(false);

  const [updateQuote] = useMutation(UpdateQuote, {
    client: b2bApolloClientContext.apolloClient,
  });

  const [createQuote] = useMutation(CreateQuote, {
    client: b2bApolloClientContext.apolloClient,
  });

  const handleCreateQuote = async ({
    currency,
    employeeId,
    employeeEmail,
    companyId,
    lineItems,
  }) => {
    if (location.state?.quote) return setQuote(location.state.quote);
    try {
      setIsLoading(true);
      const response = await createQuote({
        variables: {
          draft: {
            currency,
            employeeId,
            employeeEmail,
            companyId,
            lineItems,
          },
        },
      });
      setQuote(response.data.createQuote);
      setIsLoading(false);
    } catch (error) {
      setIsLoading(false);
      throw error;
    }
  };

  const handleUpdateQuote = async (actions, { id, version } = quote) => {
    try {
      setIsLoading(true);
      const response = await updateQuote({
        variables: {
          version,
          id,
          actions,
        },
      });
      setQuote(response.data.updateQuote);
      setIsLoading(false);
    } catch (error) {
      setIsLoading(false);
      throw error;
    }
  };

  const handleAddLineItem = async ({ quantity, sku }) =>
    handleUpdateQuote([
      {
        addLineItem: {
          sku,
          quantity,
        },
      },
    ]);

  const handleRemoveLineItem = async ({ lineItemId }) =>
    handleUpdateQuote([
      {
        removeLineItem: {
          lineItemId,
        },
      },
    ]);

  const handleChangeLineItemQuantity = async ({ lineItemId, quantity }) =>
    handleUpdateQuote([
      {
        changeLineItemQuantity: {
          lineItemId,
          quantity,
        },
      },
    ]);

  const handleConvertedQuote = ({ id, version }) => {
    handleUpdateQuote([{ changeState: { state: QUOTE_TYPES.PLACED } }], {
      id,
      version,
    });
  };

  return props.children({
    quote,
    isLoading,
    createQuote: handleCreateQuote,
    addLineItem: handleAddLineItem,
    convertedQuote: handleConvertedQuote,
    removeLineItem: handleRemoveLineItem,
    changeLineItemQuantity: handleChangeLineItemQuantity,
  });
};

QuoteCreateConnector.displayName = 'QuoteCreateConnector';
QuoteCreateConnector.propTypes = {
  children: PropTypes.func.isRequired,
};

export default QuoteCreateConnector;
