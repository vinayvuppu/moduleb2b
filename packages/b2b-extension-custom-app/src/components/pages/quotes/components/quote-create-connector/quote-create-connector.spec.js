import React from 'react';
import { shallow } from 'enzyme';

import { QuoteCreateConnector } from './quote-create-connector';

jest.mock('react-apollo', () => {
  const actual = jest.requireActual('react-apollo');
  return {
    ...actual,
    useMutation: jest.fn(() => [jest.fn().mockResolvedValue()]),
    useQuery: jest.fn(() => ({
      data: { quote: { id: 'quote-id', version: 1 }, loading: false },
      loading: false,
    })),
  };
});
jest.mock('react-router-dom', () => {
  const actual = jest.requireActual('react-router-dom');
  return { ...actual, useLocation: () => ({}) };
});

const createTestProps = props => ({
  children: jest.fn(childrenProps => <div {...childrenProps} />),
  projectKey: 'project-key-1',
  ...props,
});

describe('rendering', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<QuoteCreateConnector {...props} />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call children prop with params', () => {
    expect(props.children).toHaveBeenCalledWith({
      quote: undefined,
      isLoading: false,
      createQuote: expect.any(Function),
      addLineItem: expect.any(Function),
      removeLineItem: expect.any(Function),
      changeLineItemQuantity: expect.any(Function),
      convertedQuote: expect.any(Function),
    });
  });
});
