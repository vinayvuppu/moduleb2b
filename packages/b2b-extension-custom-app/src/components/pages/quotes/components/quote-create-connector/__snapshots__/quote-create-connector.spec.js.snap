// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should match snapshot 1`] = `
<div
  addLineItem={[Function]}
  changeLineItemQuantity={[Function]}
  convertedQuote={[Function]}
  createQuote={[Function]}
  isLoading={false}
  removeLineItem={[Function]}
/>
`;
